src/PropertyProcessor/Property/ConstProcessor.php

```diff
@@ -34,7 +34,9 @@ public function process(string $propertyName, JsonSchema $propertySchema): Prope
             $json['description'] ?? '',
         );
 
-        $property->setRequired($this->propertyMetaDataCollection->isAttributeRequired($propertyName));
+        $isAttributeRequired = $this->propertyMetaDataCollection->isAttributeRequired($propertyName);
+        $isImplicitNullAllowed = $this->schemaProcessor->getGeneratorConfiguration()->isImplicitNullAllowed();
+        $property->setRequired($isAttributeRequired || !$isImplicitNullAllowed);
```

Comment:
The required check was alright before to pull it just via `isAttributeRequired`. Now you've dropped the support for optional constants again (like eg. the ` testNotProvidedConstPropertyThrowsAnException` should not throw an exception, as the property is not required), if implicit null is not allowed. The check implementation should look like this from a quick test I think:

```php
$check = match(true) {
    $property->isRequired()
        => '$value !== ' . var_export($json['const'], true),
    $this->isImplicitNullAllowed($property)
        => '!in_array($value, ' . RenderHelper::varExportArray([$json['const'], null]) . ', true)',
    default
        => "array_key_exists('{$property->getName()}', \$modelData) && \$value !== " . var_export($json['const'], true),
};
```

Reply:
I'm so sorry, I didn't understand you at last time( Now I've made some improvements again. I think the check can be reduced to two conditions, please, take a look again

Reply:
No problem, it's not an easy library with those settings. Nice catch with the reduction of the check implementation to two conditions (although I think the version with three conditions is more suitable, as it removes an unnecessary array_key_exists check for required properties from the generated code where the check will be already executed by the `RequiredPropertyValidator`).

Reply:
We added `$modelData` to the check, I tried to generate a class from my big schema and started catching an error because $modelData is null for several nested propertis. However, I haven't find out the real reason yet

`array_key_exists(): Argument #2 ($array) must be of type array, null given`

Reply:
This might be related to a nesting of the `const` validation in a more complex validation. The more complex validations are nested in callback functions (which basically represent the same mechanic as the `$check` the const validation provides inline), where the generated code is wrapped in an if to check wether a validation is violated or not.

An example could be the `ArrayTupleValidator` which renders the `ArrayTuple.phptpl` file containing the callback function. Not all of those callback functions provide access to the `$modelData` variable and consequently a nested const validation would fail with the error you mentioned. The parent scope should always provide access to `$modelData`, consequently the access should be granted to the callback functions also (e.g. `ComposedItem.phptpl` already has the access).

This would affect the validator templates located in `src/Templates/Validator`, which make a call to `viewHelper.renderValidator`. Those callback functions, which currently don't have access to `$modelData` should get read-only access (consequently not by-reference) to enable the nested validators like the const validator to work with `$modelData`.

Reply:
Do you extend those validators? Otherwise I'd merge this MR and extend the more complex validators to also enable nested const validations (and provide some test cases as well). I've just tested it with the array tuple validation mentioned above. Given the following schema:

```json
{
  "type": "object",
  "properties": {
    "property": {
      "type": "array",
      "items": [
        {
          "const": "1"
        }
      ]
    }
  }
}
```

it will crash with the

`array_key_exists(): Argument #2 ($array) must be of type array, null given`

error without access to the `$modelData` (so the current state of the MR).

By providing it to the callback function by adding it to the scope via the use in the `ArrayTuple.phptpl` (changing the signature to `function (&$items) use (&$invalidTuples, $modelData) {`) I get the expected result.

```php
['property' => []] // works as there is no min length on the tuple array
['property' => ['1']] // works as the item doesn't violate the const validation
['property' => [1]] // fails with an InvalidTupleException "Invalid value for tuple item #0 of array property declined by const constraint"
```

Reply:
Also the same error also occurs for oneOf and etc. I found the template ComposedItem.phptpl, but didn't find the calling function template to add $modelData to the arguments.

```
{
  "type": "object",
  "properties": {
    "property": {
      "oneOf": [
        {
          "const": "1"
        },
        {
          "const": "2"
        }
      ]
    }
  }
}
```


```
...
/**
 * Execute all validators for the property property
 */
protected function validateProperty($value, array $modelData)
{
        $this->validateProperty_ComposedProperty_66eaf512f0369($value);
    return $value;
}

private function validateProperty_ComposedProperty_66eaf512f0369(&$value): void {
...
```

Reply:
Oh yeah, some more complex validations like the compositions (e.g. `oneOf`) are extracted into separate methods to be able to call them from multiple places without inserting the code (which might get pretty long if there are a lot of checks involved) multiple times.

This is done via the `ExtractedMethodValidator` which takes care to set up the extracted private method (in your example the `validateProperty_ComposedProperty_66eaf512f0369`). Consequently there the `$modelData` also needs to be added there to the generated function signature and be injected from the calling side (`RenderHelper::renderValidator`) to ensure the values are available. With this pass through I get a `OneOfException:  Invalid value for property declined by composition constraint.` when passing for example an integer value to the `property`

This is a bunch of infrastructural work to enable validators to access `$modelData` which wasn't required until now (only for the `RequiredPropertyValidator` which is not called in contexts like the `const` validator)

Reply:
Passed the variable into the templates, well, i hope it doesn't break anything else

Reply:
Looks good, I'll merge it in a separate branch, add some more test cases and then merge it and tag a new version. As the test suite is quiet large and covers a lot of cases a change should not break anything as long as the tests are green. 

Thanks for the contribution!